src/offset/local/windows.rs

```diff
@@ -8,132 +8,253 @@
+// - All other years get the base settings, which seem to be that of the current year.
+//
+// These details don't matter much, we just work with the offsets and transition dates Windows
+// returns through `GetTimeZoneInformationForYear` for a particular year.
+struct TzInfo {
+    // Offset from UTC during standard time.
+    std_offset: FixedOffset,
+    // Offset from UTC during daylight saving time.
+    dst_offset: FixedOffset,
+    // Transition from standard time to daylight saving time, given in local standard time.
+    std_transition: Option<NaiveDateTime>,
+    // Transition from daylight saving time to standard time, given in local daylight saving time.
+    dst_transition: Option<NaiveDateTime>,
 }
 
-fn from_local_time(local_time: SYSTEMTIME) -> Result<FixedOffset, Error> {
-    let utc_time = local_to_utc_time(&local_time)?;
-    let utc_secs = system_time_as_unix_seconds(&utc_time)?;
-    let local_secs = system_time_as_unix_seconds(&local_time)?;
-    let offset = (local_secs - utc_secs) as i32;
-    Ok(FixedOffset::east_opt(offset).unwrap())
+fn tz_info_for_year(year: i32) -> Option<TzInfo> {
+    // The API limits years to 1601..=30827.
+    // Working with timezones and daylight saving time this far into the past or future makes
+    // little sense. But whatever is extrapolated for 1601 or 30827 is what can be extrapolated
+    // for years beyond.
+    let ref_year = year.clamp(1601, 30827) as u16;
+    let tz_info = unsafe {
+        let mut tz_info = MaybeUninit::<TIME_ZONE_INFORMATION>::uninit();
+        if GetTimeZoneInformationForYear(ref_year, ptr::null_mut(), tz_info.as_mut_ptr()) == 0 {
+            return None;
+        }
+        tz_info.assume_init()
+    };
+    Some(TzInfo {
+        std_offset: FixedOffset::west_opt((tz_info.Bias + tz_info.StandardBias) * 60)?,
+        dst_offset: FixedOffset::west_opt((tz_info.Bias + tz_info.DaylightBias) * 60)?,
+        std_transition: systemtime_to_naive_dt(tz_info.StandardDate, year),
+        dst_transition: systemtime_to_naive_dt(tz_info.DaylightDate, year),
+    })
 }
 
-fn system_time_from_naive_date_time(dt: &NaiveDateTime) -> SYSTEMTIME {
-    SYSTEMTIME {
-        // Valid values: 1601-30827
-        wYear: dt.year() as u16,
-        // Valid values:1-12
-        wMonth: dt.month() as u16,
-        // Valid values: 0-6, starting Sunday.
-        // NOTE: enum returns 1-7, starting Monday, so we are
-        // off here, but this is not currently used in local.
-        wDayOfWeek: dt.weekday() as u16,
-        // Valid values: 1-31
-        wDay: dt.day() as u16,
-        // Valid values: 0-23
-        wHour: dt.hour() as u16,
-        // Valid values: 0-59
-        wMinute: dt.minute() as u16,
-        // Valid values: 0-59
-        wSecond: dt.second() as u16,
-        // Valid values: 0-999
-        wMilliseconds: 0,
+fn systemtime_to_naive_dt(st: SYSTEMTIME, year: i32) -> Option<NaiveDateTime> {
```

Comment:
I think I preferred the old name? (And don't see a strong reason to change it.)

Reply:
We didn't have anything like this function before. I don't mind changing the name to something else.

Reply:
Done.